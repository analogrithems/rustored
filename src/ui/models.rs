use aws_sdk_s3::primitives::DateTime as AwsDateTime;
use std::fmt;
use std::fmt::Debug;

// Re-export separated config modules
pub mod s3_config;
pub use s3_config::S3Config;

/// Restore target options
#[derive(Clone, Debug, PartialEq, Default)]
pub enum RestoreTarget {
    #[default]
    Postgres,
    Elasticsearch,
    Qdrant,
}

pub mod postgres_config;
pub use postgres_config::PostgresConfig;
pub mod elasticsearch_config;
pub use elasticsearch_config::ElasticsearchConfig;
pub mod qdrant_config;
pub use qdrant_config::QdrantConfig;

/// Input mode for the UI
#[derive(Debug, PartialEq)]
pub enum InputMode {
    Normal,
    Editing,
}

/// Metadata for a backup
#[derive(Clone, Debug, PartialEq)]
pub struct BackupMetadata {
    pub key: String,
    pub size: i64,
    pub last_modified: AwsDateTime,
}

/// State of the popup
#[derive(Debug, PartialEq)]
pub enum PopupState {
    Hidden,
    ConfirmRestore(BackupMetadata),
    Downloading(BackupMetadata, f32, f64),
    ConfirmCancel(BackupMetadata, f32, f64),
    Restoring(BackupMetadata, f32),  // Snapshot being restored, progress percentage
    TestS3Result(String),
    TestPgResult(String),
    Error(String),
    Success(String),
}

/// Focus field for the UI
#[derive(Debug, PartialEq, Clone, Copy)]
pub enum FocusField {
    // S3 Settings (10-19)
    Bucket,          // Alt+1
    Region,          // Alt+2
    Prefix,          // Alt+3
    EndpointUrl,     // Alt+4
    AccessKeyId,     // Alt+5
    SecretAccessKey, // Alt+6
    PathStyle,       // Alt+7

    // PostgreSQL Settings (20-29)
    PgHost,          // Alt+q
    PgPort,          // Alt+w
    PgUsername,      // Alt+e
    PgPassword,      // Alt+r
    PgSsl,          // Alt+t
    PgDbName,        // Alt+y
    SnapshotList,
    RestoreTarget,
    EsHost,
    EsIndex,
    QdrantApiKey,
}

impl fmt::Display for FocusField {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            FocusField::Bucket => write!(f, "Bucket"),
            FocusField::Region => write!(f, "Region"),
            FocusField::Prefix => write!(f, "Prefix"),
            FocusField::EndpointUrl => write!(f, "Endpoint URL"),
            FocusField::AccessKeyId => write!(f, "Access Key ID"),
            FocusField::SecretAccessKey => write!(f, "Secret Access Key"),
            FocusField::PathStyle => write!(f, "Path Style"),
            FocusField::SnapshotList => write!(f, "Snapshot List"),

            // Datastore Settings
            FocusField::RestoreTarget => write!(f, "Restore Target"),
            // PostgreSQL Settings (20-29)
            FocusField::PgHost => write!(f, "PostgreSQL Host"),
            FocusField::PgPort => write!(f, "PostgreSQL Port"),
            FocusField::PgUsername => write!(f, "PostgreSQL Username"),
            FocusField::PgPassword => write!(f, "PostgreSQL Password"),
            FocusField::PgSsl => write!(f, "PostgreSQL SSL"),
            FocusField::PgDbName => write!(f, "PostgreSQL Database"),
            // Elasticsearch Settings (30-39)
            FocusField::EsHost => write!(f, "Elasticsearch/Qdrant Host"),
            FocusField::EsIndex => write!(f, "Index/Collection"),
            // Qdrant Settings (40-49)
            FocusField::QdrantApiKey => write!(f, "Qdrant API Key"),
        }
    }
}
