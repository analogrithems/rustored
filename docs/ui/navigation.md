# Rustored UI Navigation Guide

This document provides a comprehensive guide to the navigation system in the Rustored terminal user interface (TUI).

## UI Layout

The Rustored UI is organized according to TDD rule #7, with a two-row layout:

```ascii
┌─────────────────┬─────────────────┬─────────────────┐
│                 │                 │                 │
│   S3 Settings   │ Restore Target  │ Target-Specific │
│                 │   Selection     │    Settings     │
│                 │                 │                 │
├─────────────────┴─────────────────┴─────────────────┤
│                                                     │
│                 Snapshot Browser                    │
│                                                     │
└─────────────────────────────────────────────────────┘
```

1. **Top Row**: Contains three panels side by side:
   - **S3 Settings** (left panel): Configuration for S3 connection parameters
   - **Restore Target Selection** (middle panel): Options to select the target datastore type (PostgreSQL, Elasticsearch, Qdrant)
   - **Target-Specific Settings** (right panel): Dynamic settings panel that changes based on the selected restore target

2. **Bottom Row**: Contains the snapshot browser that displays available snapshots from S3

For a more detailed view of the UI layout, see the [full diagram](../images/ui_layout.svg).

## Keyboard Navigation

### Window Navigation

| Key | Action |
|-----|--------|
| `Tab` | Cycle between main window sections (S3 Settings → Restore Target Settings → Snapshot List → S3 Settings) |
| `q` | Quit the application |
| `Ctrl+Z` | Suspend the application (Unix systems only) |
| `r` | Reload snapshots from S3 |
| `t` | When focus is on S3 Settings: Test S3 connection |

### Restore Target Selection

| Key | Action |
|-----|--------|
| `1` | Select PostgreSQL as the restore target |
| `2` | Select Elasticsearch as the restore target |
| `3` | Select Qdrant as the restore target |

When you select a different restore target, the Restore Settings panel will automatically update to show the appropriate settings for that target. Additionally, the focus will move to the first field in the selected target's settings if it wasn't already on a field for that target.

### Within-Window Navigation

| Key | Action |
|-----|--------|
| `↓` or `j` | Move to the next field within the current window |
| `↑` or `k` | Move to the previous field within the current window |
| `Enter` | When in Snapshot List: Select the highlighted snapshot for restore |

The up/down navigation wraps around within each window section. For example, if you're on the last field in the S3 Settings window and press Down, you'll cycle back to the first field in that window.

### Field Editing

| Key | Action |
|-----|--------|
| `Enter` | When on a field: Enter edit mode for the currently focused field |
| `Enter` | When already in edit mode: Save the edited value |
| `Esc` | Cancel editing |

## Focus Indicators

The currently focused field or section is highlighted in yellow. This provides visual feedback about which element will receive keyboard input.

## Popup Windows

Popup windows appear in certain situations:

- **Confirmation**: When confirming an action like restoring a snapshot
- **Progress**: When downloading or restoring a snapshot
- **Error/Success**: When an operation completes or fails

Press `Esc` or `Enter` to dismiss most popups.

## Example Workflow

1. Use `Tab` to navigate to the S3 Settings window
2. Use `↓`/`↑` or `j`/`k` to navigate between fields within the S3 Settings window
3. Press `Enter` to edit the currently focused field
4. Press `t` to test the S3 connection
5. Press `r` to load snapshots from S3
6. Press `Tab` to navigate to the Restore Target Settings window
7. Press `1`, `2`, or `3` to select the desired restore target
8. Use `↓`/`↑` or `j`/`k` to navigate between fields within the Restore Target Settings window
9. Press `Tab` to navigate to the Snapshot List window
10. Use `↓`/`↑` or `j`/`k` to select a snapshot
11. Press `Enter` to confirm and start the restore process

## Implementation Details

The navigation system is implemented in the `RustoredApp.handle_key_event` method, which processes all key events and updates the application state accordingly. The UI rendering is handled by the `renderer.rs` file, which uses the Ratatui library to create the layout and render the components.

When switching between restore targets, the focus is automatically updated to ensure a smooth user experience. This is achieved through the following code in the key event handler:

```rust
// Tab key handling - only changes window focus
KeyCode::Tab => {
    // Cycle between main window sections only
    self.focus = match self.focus {
        // S3 Settings fields - move to Restore Target settings
        FocusField::Bucket | FocusField::Region | /* other S3 fields */ => {
            // Move to restore target settings
            match self.restore_target {
                RestoreTarget::Postgres => FocusField::PgHost,
                RestoreTarget::Elasticsearch => FocusField::EsHost,
                RestoreTarget::Qdrant => FocusField::EsHost,
            }
        }
        // Restore Target settings - move to Snapshot List
        FocusField::PgHost | /* other Postgres fields */ |
        FocusField::EsHost | /* other ES fields */ => FocusField::SnapshotList,
        // Snapshot list - move back to S3 Settings
        FocusField::SnapshotList => FocusField::Bucket,
        // Default case
        _ => FocusField::Bucket,
    };
}

// Down key handling - navigates within a window
KeyCode::Down | KeyCode::Char('j') => {
    if self.focus == FocusField::SnapshotList {
        // Navigate snapshot list
        self.snapshot_browser.selected_index = 
            (self.snapshot_browser.selected_index + 1) % self.snapshot_browser.snapshots.len();
    } else {
        // Navigate within settings windows
        self.focus = match self.focus {
            // S3 Settings navigation
            FocusField::Bucket => FocusField::Region,
            // ... other field navigation logic
        };
    }
}
