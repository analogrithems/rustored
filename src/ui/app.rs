use std::io::stdout;
use log::debug;
use std::time::Duration;
use anyhow::Result;
use crossterm::event::{self, Event, KeyCode, KeyModifiers};
use crossterm::terminal::{disable_raw_mode, enable_raw_mode};
use crossterm::execute;
use crossterm::terminal::{EnterAlternateScreen, LeaveAlternateScreen};
use crossterm::event::DisableMouseCapture;
use crossterm::event::EnableMouseCapture;
use libc::SIGTSTP;
use libc::raise;
use ratatui::backend::Backend;
use ratatui::Terminal;

use crate::ui::models::InputMode;
use crate::ui::rustored::RustoredApp;

/// Run the TUI application, delegating to RustoredApp
pub async fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut RustoredApp) -> Result<Option<String>> {
    // Initial load of snapshots
    if let Err(e) = app.snapshot_browser.load_snapshots().await {
        debug!("Failed to load snapshots: {}", e);
    }

    loop {
        // Draw UI
        terminal.draw(|f| crate::ui::renderer::ui::<B>(f, app))?;

        // Handle events
        if event::poll(Duration::from_millis(100))? {
            if let Event::Key(key) = event::read()? {
                // Handle suspend on Ctrl+Z when in Normal mode
                if app.input_mode == InputMode::Normal && key.modifiers.contains(KeyModifiers::CONTROL) && key.code == KeyCode::Char('z') {
                    // Exit TUI mode
                    disable_raw_mode()?;
                    execute!(stdout(), LeaveAlternateScreen, DisableMouseCapture)?;
                    // Suspend process
                    unsafe { raise(SIGTSTP); }
                    // On resume, re-enter TUI
                    enable_raw_mode()?;
                    execute!(stdout(), EnterAlternateScreen, EnableMouseCapture)?;
                    continue;
                }
                
                // Pass the key event to the app
                if let Some(snapshot_path) = app.handle_key_event::<B>(key).await? {
                    return Ok(Some(snapshot_path));
                }
            }
        }
    }
}
