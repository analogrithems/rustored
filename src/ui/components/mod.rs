pub mod s3_settings;
pub mod restore_target;
pub mod snapshot_list;
pub mod popups;

// Re-export all components for easier imports
pub use s3_settings::*;
pub use restore_target::*;
pub use snapshot_list::*;
pub use popups::*;
