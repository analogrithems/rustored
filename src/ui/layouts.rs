use ratatui::layout::{Constraint, Direction, Layout, Rect};
use log::debug;

/// Helper function to create a centered rect
pub fn centered_rect(percent_x: u16, height: u16, r: Rect) -> Rect {
    debug!("Creating centered rect with percent_x: {}, height: {}, in area: {:?}", percent_x, height, r);
    let popup_layout = Layout::default()
        .direction(Direction::Vertical)
        .constraints(
            [
                Constraint::Percentage((100 - height) / 2),
                Constraint::Length(height),
                Constraint::Percentage((100 - height) / 2),
            ]
            .as_ref(),
        )
        .split(r);

    let result = Layout::default()
        .direction(Direction::Horizontal)
        .constraints(
            [
                Constraint::Percentage((100 - percent_x) / 2),
                Constraint::Percentage(percent_x),
                Constraint::Percentage((100 - percent_x) / 2),
            ]
            .as_ref(),
        )
        .split(popup_layout[1])[1];
    
    debug!("Created centered rect: {:?}", result);
    result
}
