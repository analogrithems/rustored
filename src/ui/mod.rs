// UI module for the postgres manager
pub mod models;
pub mod browser;
pub mod renderer;

// Re-export the main entry point
pub use browser::run_tui;
