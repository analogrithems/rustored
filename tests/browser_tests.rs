use insta::assert_debug_snapshot;

// We need to use the crate name directly since we're in an integration test
use rustored::ui::models::{BackupMetadata, FocusField, InputMode, PopupState, PostgresConfig, S3Config, ElasticsearchConfig, QdrantConfig, RestoreTarget};

// Helper function to create a test timestamp with a fixed value
fn create_test_aws_datetime() -> f64 {
    // Use a fixed timestamp for testing (2023-01-01 12:00:00 UTC)
    1672574400.0
}

#[test]
fn test_s3_config() {
    let s3_config = S3Config {
        bucket: "test-bucket".to_string(),
        region: "us-west-2".to_string(),
        prefix: "test-prefix".to_string(),
        endpoint_url: "https://test-endpoint.com".to_string(),
        access_key_id: "test-access-key".to_string(),
        secret_access_key: "test-secret-key".to_string(),
        path_style: false,
        error_message: None,
        test_s3_button: false,
    };

    assert_debug_snapshot!(s3_config);
}

#[test]
fn test_s3_config_focus_fields() {
    // Test that we can get all focus fields for S3 settings
    let fields = S3Config::focus_fields();
    
    // Verify we have the expected number of fields
    assert_eq!(fields.len(), 7);
    
    // Verify all expected fields are present
    assert!(fields.contains(&FocusField::Bucket));
    assert!(fields.contains(&FocusField::Region));
    assert!(fields.contains(&FocusField::Prefix));
    assert!(fields.contains(&FocusField::EndpointUrl));
    assert!(fields.contains(&FocusField::AccessKeyId));
    assert!(fields.contains(&FocusField::SecretAccessKey));
    assert!(fields.contains(&FocusField::PathStyle));
}

#[test]
fn test_s3_config_contains_field() {
    // Test that contains_field correctly identifies S3 fields
    assert!(S3Config::contains_field(FocusField::Bucket));
    assert!(S3Config::contains_field(FocusField::Region));
    assert!(S3Config::contains_field(FocusField::Prefix));
    assert!(S3Config::contains_field(FocusField::EndpointUrl));
    assert!(S3Config::contains_field(FocusField::AccessKeyId));
    assert!(S3Config::contains_field(FocusField::SecretAccessKey));
    assert!(S3Config::contains_field(FocusField::PathStyle));
    
    // Test that it correctly rejects non-S3 fields
    assert!(!S3Config::contains_field(FocusField::PgHost));
    assert!(!S3Config::contains_field(FocusField::SnapshotList));
    assert!(!S3Config::contains_field(FocusField::RestoreTarget));
}

#[test]
fn test_s3_config_get_field_value() {
    let s3_config = S3Config {
        bucket: "test-bucket".to_string(),
        region: "us-west-2".to_string(),
        prefix: "test-prefix".to_string(),
        endpoint_url: "https://test-endpoint.com".to_string(),
        access_key_id: "test-access-key".to_string(),
        secret_access_key: "test-secret-key".to_string(),
        path_style: true,
        error_message: None,
        test_s3_button: false,
    };
    
    // Test getting field values
    assert_eq!(s3_config.get_field_value(FocusField::Bucket), "test-bucket");
    assert_eq!(s3_config.get_field_value(FocusField::Region), "us-west-2");
    assert_eq!(s3_config.get_field_value(FocusField::Prefix), "test-prefix");
    assert_eq!(s3_config.get_field_value(FocusField::EndpointUrl), "https://test-endpoint.com");
    assert_eq!(s3_config.get_field_value(FocusField::AccessKeyId), "test-access-key");
    assert_eq!(s3_config.get_field_value(FocusField::SecretAccessKey), "test-secret-key");
    assert_eq!(s3_config.get_field_value(FocusField::PathStyle), "true");
    
    // Test getting a non-S3 field (should return empty string)
    assert_eq!(s3_config.get_field_value(FocusField::PgHost), "");
}

#[test]
fn test_s3_config_set_field_value() {
    let mut s3_config = S3Config {
        bucket: "".to_string(),
        region: "".to_string(),
        prefix: "".to_string(),
        endpoint_url: "".to_string(),
        access_key_id: "".to_string(),
        secret_access_key: "".to_string(),
        path_style: false,
        error_message: None,
        test_s3_button: false,
    };
    
    // Test setting field values
    s3_config.set_field_value(FocusField::Bucket, "new-bucket".to_string());
    s3_config.set_field_value(FocusField::Region, "eu-west-1".to_string());
    s3_config.set_field_value(FocusField::Prefix, "new-prefix".to_string());
    s3_config.set_field_value(FocusField::EndpointUrl, "https://new-endpoint.com".to_string());
    s3_config.set_field_value(FocusField::AccessKeyId, "new-access-key".to_string());
    s3_config.set_field_value(FocusField::SecretAccessKey, "new-secret-key".to_string());
    s3_config.set_field_value(FocusField::PathStyle, "true".to_string());
    
    // Verify the values were set correctly
    assert_eq!(s3_config.bucket, "new-bucket");
    assert_eq!(s3_config.region, "eu-west-1");
    assert_eq!(s3_config.prefix, "new-prefix");
    assert_eq!(s3_config.endpoint_url, "https://new-endpoint.com");
    assert_eq!(s3_config.access_key_id, "new-access-key");
    assert_eq!(s3_config.secret_access_key, "new-secret-key");
    assert_eq!(s3_config.path_style, true);
    
    // Test setting a non-S3 field (should have no effect)
    s3_config.set_field_value(FocusField::PgHost, "should-not-change-anything".to_string());
    assert_eq!(s3_config.bucket, "new-bucket"); // Verify no change
}

#[test]
fn test_postgres_config() {
    let pg_config = PostgresConfig {
        host: Some("localhost".to_string()),
        port: Some(5432),
        username: Some("postgres".to_string()),
        password: Some("password".to_string()),
        use_ssl: false,
        db_name: Some("postgres".to_string()),
    };

    assert_debug_snapshot!(pg_config);
}

#[test]
fn test_postgres_config_focus_fields() {
    // Test that we can get all focus fields for PostgreSQL settings
    let fields = PostgresConfig::focus_fields();
    
    // Verify we have the expected number of fields
    assert_eq!(fields.len(), 6);
    
    // Verify all expected fields are present
    assert!(fields.contains(&FocusField::PgHost));
    assert!(fields.contains(&FocusField::PgPort));
    assert!(fields.contains(&FocusField::PgUsername));
    assert!(fields.contains(&FocusField::PgPassword));
    assert!(fields.contains(&FocusField::PgSsl));
    assert!(fields.contains(&FocusField::PgDbName));
}

#[test]
fn test_postgres_config_contains_field() {
    // Test that contains_field correctly identifies PostgreSQL fields
    assert!(PostgresConfig::contains_field(FocusField::PgHost));
    assert!(PostgresConfig::contains_field(FocusField::PgPort));
    assert!(PostgresConfig::contains_field(FocusField::PgUsername));
    assert!(PostgresConfig::contains_field(FocusField::PgPassword));
    assert!(PostgresConfig::contains_field(FocusField::PgSsl));
    assert!(PostgresConfig::contains_field(FocusField::PgDbName));
    
    // Test that it correctly rejects non-PostgreSQL fields
    assert!(!PostgresConfig::contains_field(FocusField::Bucket));
    assert!(!PostgresConfig::contains_field(FocusField::SnapshotList));
    assert!(!PostgresConfig::contains_field(FocusField::RestoreTarget));
}

#[test]
fn test_postgres_config_get_field_value() {
    let pg_config = PostgresConfig {
        host: Some("localhost".to_string()),
        port: Some(5432),
        username: Some("postgres".to_string()),
        password: Some("password".to_string()),
        use_ssl: true,
        db_name: Some("postgres".to_string()),
    };
    
    // Test getting field values
    assert_eq!(pg_config.get_field_value(FocusField::PgHost), "localhost");
    assert_eq!(pg_config.get_field_value(FocusField::PgPort), "5432");
    assert_eq!(pg_config.get_field_value(FocusField::PgUsername), "postgres");
    assert_eq!(pg_config.get_field_value(FocusField::PgPassword), "password");
    assert_eq!(pg_config.get_field_value(FocusField::PgSsl), "true");
    assert_eq!(pg_config.get_field_value(FocusField::PgDbName), "postgres");
    
    // Test getting a non-PostgreSQL field (should return empty string)
    assert_eq!(pg_config.get_field_value(FocusField::Bucket), "");
    
    // Test with None values
    let empty_pg_config = PostgresConfig {
        host: None,
        port: None,
        username: None,
        password: None,
        use_ssl: false,
        db_name: None,
    };
    
    assert_eq!(empty_pg_config.get_field_value(FocusField::PgHost), "");
    assert_eq!(empty_pg_config.get_field_value(FocusField::PgPort), "");
    assert_eq!(empty_pg_config.get_field_value(FocusField::PgUsername), "");
    assert_eq!(empty_pg_config.get_field_value(FocusField::PgPassword), "");
    assert_eq!(empty_pg_config.get_field_value(FocusField::PgSsl), "false");
    assert_eq!(empty_pg_config.get_field_value(FocusField::PgDbName), "");
}

#[test]
fn test_postgres_config_set_field_value() {
    let mut pg_config = PostgresConfig {
        host: None,
        port: None,
        username: None,
        password: None,
        use_ssl: false,
        db_name: None,
    };
    
    // Test setting field values
    pg_config.set_field_value(FocusField::PgHost, "new-host".to_string());
    pg_config.set_field_value(FocusField::PgPort, "5433".to_string());
    pg_config.set_field_value(FocusField::PgUsername, "new-user".to_string());
    pg_config.set_field_value(FocusField::PgPassword, "new-password".to_string());
    pg_config.set_field_value(FocusField::PgSsl, "true".to_string());
    pg_config.set_field_value(FocusField::PgDbName, "new-database".to_string());
    
    // Verify the values were set correctly
    assert_eq!(pg_config.host, Some("new-host".to_string()));
    assert_eq!(pg_config.port, Some(5433));
    assert_eq!(pg_config.username, Some("new-user".to_string()));
    assert_eq!(pg_config.password, Some("new-password".to_string()));
    assert_eq!(pg_config.use_ssl, true);
    assert_eq!(pg_config.db_name, Some("new-database".to_string()));
    
    // Test setting a non-PostgreSQL field (should have no effect)
    pg_config.set_field_value(FocusField::Bucket, "should-not-change-anything".to_string());
    assert_eq!(pg_config.host, Some("new-host".to_string())); // Verify no change
    
    // Test with invalid port value
    pg_config.set_field_value(FocusField::PgPort, "not-a-number".to_string());
    assert_eq!(pg_config.port, None); // Should be None when parse fails
}

#[test]
fn test_elasticsearch_config() {
    let es_config = ElasticsearchConfig {
        host: Some("http://localhost:9200".to_string()),
        index: Some("test-index".to_string()),
    };

    assert_debug_snapshot!(es_config);
}

#[test]
fn test_elasticsearch_config_focus_fields() {
    // Test that we can get all focus fields for Elasticsearch settings
    let fields = ElasticsearchConfig::focus_fields();
    
    // Verify we have the expected number of fields
    assert_eq!(fields.len(), 2);
    
    // Verify all expected fields are present
    assert!(fields.contains(&FocusField::EsHost));
    assert!(fields.contains(&FocusField::EsIndex));
}

#[test]
fn test_elasticsearch_config_contains_field() {
    // Test that contains_field correctly identifies Elasticsearch fields
    assert!(ElasticsearchConfig::contains_field(FocusField::EsHost));
    assert!(ElasticsearchConfig::contains_field(FocusField::EsIndex));
    
    // Test that it correctly rejects non-Elasticsearch fields
    assert!(!ElasticsearchConfig::contains_field(FocusField::Bucket));
    assert!(!ElasticsearchConfig::contains_field(FocusField::PgHost));
    assert!(!ElasticsearchConfig::contains_field(FocusField::QdrantApiKey));
}

#[test]
fn test_elasticsearch_config_get_field_value() {
    let es_config = ElasticsearchConfig {
        host: Some("http://localhost:9200".to_string()),
        index: Some("test-index".to_string()),
    };
    
    // Test getting field values
    assert_eq!(es_config.get_field_value(FocusField::EsHost), "http://localhost:9200");
    assert_eq!(es_config.get_field_value(FocusField::EsIndex), "test-index");
    
    // Test getting a non-Elasticsearch field (should return empty string)
    assert_eq!(es_config.get_field_value(FocusField::Bucket), "");
    
    // Test with None values
    let empty_es_config = ElasticsearchConfig {
        host: None,
        index: None,
    };
    
    assert_eq!(empty_es_config.get_field_value(FocusField::EsHost), "");
    assert_eq!(empty_es_config.get_field_value(FocusField::EsIndex), "");
}

#[test]
fn test_elasticsearch_config_set_field_value() {
    let mut es_config = ElasticsearchConfig {
        host: None,
        index: None,
    };
    
    // Test setting field values
    es_config.set_field_value(FocusField::EsHost, "http://new-host:9200".to_string());
    es_config.set_field_value(FocusField::EsIndex, "new-index".to_string());
    
    // Verify the values were set correctly
    assert_eq!(es_config.host, Some("http://new-host:9200".to_string()));
    assert_eq!(es_config.index, Some("new-index".to_string()));
    
    // Test setting a non-Elasticsearch field (should have no effect)
    es_config.set_field_value(FocusField::Bucket, "should-not-change-anything".to_string());
    assert_eq!(es_config.host, Some("http://new-host:9200".to_string())); // Verify no change
}

#[test]
fn test_qdrant_config() {
    let qdrant_config = QdrantConfig {
        host: Some("http://localhost:6333".to_string()),
        collection: Some("test-collection".to_string()),
        api_key: Some("test-api-key".to_string()),
    };

    assert_debug_snapshot!(qdrant_config);
}

#[test]
fn test_qdrant_config_focus_fields() {
    // Test that we can get all focus fields for Qdrant settings
    let fields = QdrantConfig::focus_fields();
    
    // Verify we have the expected number of fields
    assert_eq!(fields.len(), 3);
    
    // Verify all expected fields are present
    assert!(fields.contains(&FocusField::EsHost)); // Reused for Qdrant host
    assert!(fields.contains(&FocusField::EsIndex)); // Reused for collection
    assert!(fields.contains(&FocusField::QdrantApiKey));
}

#[test]
fn test_qdrant_config_contains_field() {
    // Test that contains_field correctly identifies Qdrant fields
    assert!(QdrantConfig::contains_field(FocusField::EsHost)); // Reused for Qdrant host
    assert!(QdrantConfig::contains_field(FocusField::EsIndex)); // Reused for collection
    assert!(QdrantConfig::contains_field(FocusField::QdrantApiKey));
    
    // Test that it correctly rejects non-Qdrant fields
    assert!(!QdrantConfig::contains_field(FocusField::Bucket));
    assert!(!QdrantConfig::contains_field(FocusField::PgHost));
}

#[test]
fn test_qdrant_config_get_field_value() {
    let qdrant_config = QdrantConfig {
        host: Some("http://localhost:6333".to_string()),
        collection: Some("test-collection".to_string()),
        api_key: Some("test-api-key".to_string()),
    };
    
    // Test getting field values
    assert_eq!(qdrant_config.get_field_value(FocusField::EsHost), "http://localhost:6333");
    assert_eq!(qdrant_config.get_field_value(FocusField::EsIndex), "test-collection");
    assert_eq!(qdrant_config.get_field_value(FocusField::QdrantApiKey), "test-api-key");
    
    // Test getting a non-Qdrant field (should return empty string)
    assert_eq!(qdrant_config.get_field_value(FocusField::Bucket), "");
    
    // Test with None values
    let empty_qdrant_config = QdrantConfig {
        host: None,
        collection: None,
        api_key: None,
    };
    
    assert_eq!(empty_qdrant_config.get_field_value(FocusField::EsHost), "");
    assert_eq!(empty_qdrant_config.get_field_value(FocusField::EsIndex), "");
    assert_eq!(empty_qdrant_config.get_field_value(FocusField::QdrantApiKey), "");
}

#[test]
fn test_qdrant_config_set_field_value() {
    let mut qdrant_config = QdrantConfig {
        host: None,
        collection: None,
        api_key: None,
    };
    
    // Test setting field values
    qdrant_config.set_field_value(FocusField::EsHost, "http://new-host:6333".to_string());
    qdrant_config.set_field_value(FocusField::EsIndex, "new-collection".to_string());
    qdrant_config.set_field_value(FocusField::QdrantApiKey, "new-api-key".to_string());
    
    // Verify the values were set correctly
    assert_eq!(qdrant_config.host, Some("http://new-host:6333".to_string()));
    assert_eq!(qdrant_config.collection, Some("new-collection".to_string()));
    assert_eq!(qdrant_config.api_key, Some("new-api-key".to_string()));
    
    // Test setting a non-Qdrant field (should have no effect)
    qdrant_config.set_field_value(FocusField::Bucket, "should-not-change-anything".to_string());
    assert_eq!(qdrant_config.host, Some("http://new-host:6333".to_string())); // Verify no change
}

#[test]
fn test_backup_metadata() {
    let datetime = create_test_aws_datetime();
    let backup = BackupMetadata {
        key: "test-snapshot-1.sql.gz".to_string(),
        size: 1024 * 1024 * 10, // 10 MB
        last_modified: datetime,
    };

    assert_debug_snapshot!(backup);
}

#[test]
fn test_popup_states() {
    let datetime = create_test_aws_datetime();
    let backup = BackupMetadata {
        key: "test-snapshot-1.sql.gz".to_string(),
        size: 1024 * 1024 * 10, // 10 MB
        last_modified: datetime,
    };

    let hidden = PopupState::Hidden;
    let confirm_restore = PopupState::ConfirmRestore(backup.clone());
    let downloading = PopupState::Downloading(backup.clone(), 0.5, 1024.0 * 1024.0);
    let confirm_cancel = PopupState::ConfirmCancel(backup.clone(), 0.5, 1024.0 * 1024.0);
    let error = PopupState::Error("Test error message".to_string());
    let success = PopupState::Success("Test success message".to_string());

    assert_debug_snapshot!("popup_state_hidden", hidden);
    assert_debug_snapshot!("popup_state_confirm_restore", confirm_restore);
    assert_debug_snapshot!("popup_state_downloading", downloading);
    assert_debug_snapshot!("popup_state_confirm_cancel", confirm_cancel);
    assert_debug_snapshot!("popup_state_error", error);
    assert_debug_snapshot!("popup_state_success", success);
}

#[test]
fn test_focus_fields() {
    assert_debug_snapshot!("focus_field_snapshot_list", FocusField::SnapshotList);
    assert_debug_snapshot!("focus_field_bucket", FocusField::Bucket);
    assert_debug_snapshot!("focus_field_region", FocusField::Region);
    assert_debug_snapshot!("focus_field_prefix", FocusField::Prefix);
    assert_debug_snapshot!("focus_field_endpoint_url", FocusField::EndpointUrl);
    assert_debug_snapshot!("focus_field_access_key_id", FocusField::AccessKeyId);
    assert_debug_snapshot!("focus_field_secret_access_key", FocusField::SecretAccessKey);
    assert_debug_snapshot!("focus_field_path_style", FocusField::PathStyle);
}

#[test]
fn test_restore_target() {
    assert_debug_snapshot!("restore_target_postgres", RestoreTarget::Postgres);
    assert_debug_snapshot!("restore_target_elasticsearch", RestoreTarget::Elasticsearch);
    assert_debug_snapshot!("restore_target_qdrant", RestoreTarget::Qdrant);
}

#[test]
fn test_restore_target_focus_fields() {
    // Test that we can get focus fields for each restore target
    let postgres_fields = RestoreTarget::Postgres.focus_fields();
    let elasticsearch_fields = RestoreTarget::Elasticsearch.focus_fields();
    let qdrant_fields = RestoreTarget::Qdrant.focus_fields();
    
    // Verify we have the expected number of fields for each target
    assert_eq!(postgres_fields.len(), 6);
    assert_eq!(elasticsearch_fields.len(), 2);
    assert_eq!(qdrant_fields.len(), 3);
    
    // Verify first field for each target
    assert_eq!(RestoreTarget::Postgres.first_focus_field(), FocusField::PgHost);
    assert_eq!(RestoreTarget::Elasticsearch.first_focus_field(), FocusField::EsHost);
    assert_eq!(RestoreTarget::Qdrant.first_focus_field(), FocusField::QdrantApiKey);
    
    // Verify fields for Postgres
    assert!(postgres_fields.contains(&FocusField::PgHost));
    assert!(postgres_fields.contains(&FocusField::PgPort));
    assert!(postgres_fields.contains(&FocusField::PgUsername));
    assert!(postgres_fields.contains(&FocusField::PgPassword));
    assert!(postgres_fields.contains(&FocusField::PgSsl));
    assert!(postgres_fields.contains(&FocusField::PgDbName));
    
    // Verify fields for Elasticsearch
    assert!(elasticsearch_fields.contains(&FocusField::EsHost));
    assert!(elasticsearch_fields.contains(&FocusField::EsIndex));
    
    // Verify fields for Qdrant
    assert!(qdrant_fields.contains(&FocusField::EsHost));
    assert!(qdrant_fields.contains(&FocusField::EsIndex));
    assert!(qdrant_fields.contains(&FocusField::QdrantApiKey));
}

#[test]
fn test_input_modes() {
    assert_debug_snapshot!("input_mode_normal", InputMode::Normal);
    assert_debug_snapshot!("input_mode_editing", InputMode::Editing);
}
