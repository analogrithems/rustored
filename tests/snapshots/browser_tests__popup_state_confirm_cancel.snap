---
source: tests/browser_tests.rs
expression: confirm_cancel
---
ConfirmCancel(
    BackupMetadata {
        key: "test-snapshot-1.sql.gz",
        size: 10485760,
        last_modified: 1672574400.0,
    },
    0.5,
    1048576.0,
)
